plugins/genesys/fu-genesys-hubhid-device.c

```diff
@@ -0,0 +1,351 @@
+					    buf_report->len,
+					    0, /* dst */
+					    fu_chunk_get_data(chk),
+					    fu_chunk_get_data_sz(chk),
+					    0x0, /* src */
+					    fu_chunk_get_data_sz(chk),
+					    error)) {
+				g_prefix_error(error,
+					       "error setting report data at 0x%04x: ",
+					       fu_chunk_get_address(chk));
+				return FALSE;
+			}
+			if (!fu_hid_device_set_report(hid_device,
+						      GENESYS_HUBHID_REPORT_ID,
+						      buf_report->data,
+						      buf_report->len,
+						      GENESYS_HUBHID_REPORT_TIMEOUT,
+						      GENESYS_HUBHID_REPORT_FLAGS |
+							  FU_HID_DEVICE_FLAG_RETRY_FAILURE,
+						      error)) {
+				g_prefix_error(error,
+					       "error setting report at 0x%04x: ",
+					       fu_chunk_get_address(chk));
+				return FALSE;
+			}
+			if (progress != NULL)
+				fu_progress_step_done(progress);
+		}
+	}
+
+	/* finish report */
+	if (!fu_hid_device_get_report(hid_device,
+				      GENESYS_HUBHID_REPORT_ID,
+				      buf_report->data,
+				      buf_report->len,
+				      GENESYS_HUBHID_REPORT_TIMEOUT,
+				      GENESYS_HUBHID_REPORT_FLAGS,
+				      error)) {
+		g_prefix_error(error, "error finishing report: ");
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_genesys_hubhid_device_validate_token(FuGenesysHubhidDevice *self, GError **error)
+{
+	const guint16 hid_token[] = {
+	    GUINT16_TO_LE('G'),
+	    GUINT16_TO_LE('L'),
+	    GUINT16_TO_LE('I'),
+	    GUINT16_TO_LE(' '),
+	    GUINT16_TO_LE('H'),
+	    GUINT16_TO_LE('I'),
+	    GUINT16_TO_LE('D'),
+	};
+	g_autoptr(GByteArray) buf_data = g_byte_array_new();
+	g_autofree FuGenesysUsbSetup *setup = g_new0(FuGenesysUsbSetup, 1);
+
+	/* get 0x80 string descriptor */
+	setup->req_type.bm = 0x80;
+	setup->request = 0x06;
+	setup->value = (0x03 << 8) | 0x80;
+	setup->index = 0;
+	setup->length = 0x40;
+
+	fu_byte_array_set_size(buf_data, setup->length, 0);
+
+	if (!fu_genesys_hubhid_device_command_read(self, setup, buf_data->data, NULL, error))
+		return FALSE;
+	if (!fu_memcmp_safe(buf_data->data,
+			    buf_data->len,
+			    2,
+			    (const guint8 *)hid_token,
+			    sizeof(hid_token),
+			    0,
+			    sizeof(hid_token),
+			    error)) {
+		g_prefix_error(error, "wrong HID token string: ");
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_genesys_hubhid_device_probe(FuDevice *device, GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev(FU_USB_DEVICE(device));
+
+	if (g_usb_device_get_device_class(usb_device) != G_USB_DEVICE_CLASS_INTERFACE_DESC) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_SUPPORTED,
+				    "is not a hub hid");
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_genesys_hubhid_device_open(FuDevice *device, GError **error)
+{
+	/* FuHidDevice->open */
+	if (!FU_DEVICE_CLASS(fu_genesys_hubhid_device_parent_class)->open(device, error))
+		return FALSE;
+
+	return TRUE;
+}
```

Comment:
do we need this?

Reply:
The `hubhid` is a command channel for hub, and we keep operating to hub.
I have tried using proxy for hub first, but I misunderstand the proxy mechanism.
Therefore, I add `->open()`/`->close()` in `usbhub` and `hubhid` to open the hid device.
https://github.com/fwupd/fwupd/blob/dc5c837ece66c64ff009b5e7b840b85b61b413ac/plugins/genesys/fu-genesys-usbhub-device.c#L1397-L1431

Reply:
Right, but I think the code you've got above is a NOP -- i.e. if you don't override FuGenesysHubhidDevice->open then it'll fall back to FuHidDevice->open - i.e. you could delete the ->open function and it'd work the same as now.